migrations/main.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"badi-cal/badi-events-backend/orm"
+	"badi-cal/badi-events-backend/terminal"
+	"fmt"
+	"log"
+
+	"gorm.io/gorm"
+)
+
+type SchemaMigration struct {
+	Migration string `gorm:"primaryKey"`
+}
+
+type MigrationAction func(db *gorm.DB)
+type Migration struct {
+	version string
+	up      MigrationAction
+	down    MigrationAction
+}
+
+// var argLength int = -1
+var versions []Migration
+
+func (migration *Migration) Migrate() {
+	db := orm.Connection()
+	var schema_migration SchemaMigration
+	if result := db.Where("migration = ?", migration.version).Take(&schema_migration); result.Error != nil {
+		fmt.Printf("Running migration '%s'", migration.version)
+		migration.up(db)
+		schema_migration = SchemaMigration{Migration: migration.version}
+		result := db.Create(&schema_migration)
+		if result.Error != nil {
+			log.Fatalf("Unable to mark migration. %v\n", result.Error)
+		}
+	} else {
+		fmt.Printf("Skipping migration '%s'; already run.\n", migration.version)
+		fmt.Printf("Error %v\n", result.Error)
+	}
+}
+
+func (migration *Migration) Rollback() {
```

Comment:
I find this `Rollback` function interesting.  A long time ago I learned that you can never really rollback a database migration.  For instance if part of the migration was to remove a column of a table or remove an entire table of a database then you can never really rollback from that as the data that either that column table or database table had are pretty much gone.  I guess this could be good to have, but I would still be interested in seeing it being used.

Reply:
I find it useful to test migrations. I've heard the argument for not having rollbacks, and I agree with it but only for prod. I think its important to be able to reset things easily for testing, etc.

Reply:
That makes sense.  Sounds good.